### ISSUE TITLE
Autocorrection of Lint/LiteralInInterpolation changes code behaviour in regexes

### ISSUE THREAD

**Issue by patrick-gleeson** [rubocop-hq/rubocop#6822]
Rubocop generally auto-replaces `"#{'a'}"` with `"a"` as a response to the "Lint/LiteralInInterpolation: Literal interpolation detected." offence. However, when it comes to regex capture group literals, this changes the behaviour of the code, making this auto-correction unsafe

--------

## Expected behavior

When running `rubocop -a` on this line of code:

```ruby
x = "ABC".gsub(/(A)(B)(C)/, "D#{'\2'}F")
```

The resulting code should look like:

```ruby
x = "ABC".gsub(/(A)(B)(C)/, "D\\2F")
```

In both cases, the value of `x` is `"DBF"`

## Actual behavior

When running `rubocop -a` on this line of code:

```ruby
x = "ABC".gsub(/(A)(B)(C)/, "D#{'\2'}F")
```

The resulting code actually looks like:

```ruby
x = "ABC".gsub(/(A)(B)(C)/, "D\2F")
```

The value of `x` is now `"D\u0002F"`

## Steps to reproduce the problem

Put this in a file:

```ruby
"ABC".gsub(/(A)(B)(C)/, "D#{'\2'}F")
```
run `rubocop -a /path/to/file.rb`

## RuboCop version

```
$ [bundle exec] rubocop -V
0.65.0 (using Parser 2.6.0.0, running on ruby 2.5.3 x86_64-darwin17)
```

**Comment by rrosenblum** [comment]
Does anyone have thoughts on the best way to fix this? Ignore the offenses if they are an argument to `sub`, `gsub`, and similar methods? Do not register an offense if the string contains `\<number>`?

**Comment by hoshinotsuyoshi** [comment]
@rrosenblum 

I think that there is no need to treat `#gsub`, any methods, or `\<number>` specially.

The problem is that auto-correcting results in wrong value replacing; like `"\\2"` -> `"\2"`.

---

I think that [RuboCop::Cop::Util#escape_string](https://github.com/rubocop-hq/rubocop/blob/v0.67.2/lib/rubocop/cop/util.rb#L76-L78) will fit to fix that...

```diff
--- a/lib/rubocop/cop/lint/literal_in_interpolation.rb
+++ b/lib/rubocop/cop/lint/literal_in_interpolation.rb
@@ -55,7 +55,7 @@ module RuboCop
           when :float
             node.children.last.to_f.to_s
           when :str
-            node.children.last
+            escape_string(node.children.last)
           when :sym
             autocorrected_value_for_symbol(node)
           else
```

I will investigate later.

Escaping is difficult...

**Comment by hoshinotsuyoshi** [comment]
> I think that RuboCop::Cop::Util#escape_string will fit to fix that...

Sorry, I found that it breaks other behaviors.  😭

Forget that...

<details><summary>memo:</summary>


```ruby
"foo#{"\n"}bar"
```

now latest rubocop (v0.67.2) auto-corrects to that

```ruby
"foo
bar"
```

but using `RuboCop::Cop::Util#escape_string` breaks this:

```ruby
"foo\nbar"
```


</details>

**Comment by hoshinotsuyoshi** [comment]
But I believe that the current behavior below is simply a bug.

```ruby
# a.rb
"foo#{'\n'}bar"
```

rubocop (v0.67.2) corrects

```
rubocop -a a.rb --only Lint/LiteralInInterpolation
Inspecting 1 file
W

Offenses:

a.rb:3:7: W: [Corrected] Lint/LiteralInInterpolation: Literal interpolation detected.
"foo#{'\n'}bar"
      ^^^^

1 file inspected, 1 offense detected, 1 offense corrected
```

results in:

```ruby
# a.rb
"foo\nbar"
```

**Comment by patrick-gleeson** [comment]
@hoshinotsuyoshi you are a true pillar of the community. Thank you for fixing this!
